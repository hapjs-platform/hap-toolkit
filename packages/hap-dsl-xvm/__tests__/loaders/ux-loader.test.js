/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

const path = require('path')
const { rowify } = require('hap-dev-utils')
const uxLoader = require('../../lib/loaders/ux-loader').default
const { compileOptionsObject } = require('@hap-toolkit/shared-utils')

describe('ux-loader', () => {
  const mockLoader = {
    context: __dirname,
    rootContext: __dirname
  }
  const samples = [
    [
      'normal page',
      `
    <template>
      <div>{{ val }}</div>
    </template>
      `,
      'page',
      'page.ux'
    ],
    [
      'normal page with style',
      `
      <template>
        <div>{{ val }}</div>
      </template>
      <style>
      div {
        color: blue;
      }
      </style>
      `,
      'page',
      'page.ux'
    ],
    [
      'normal page, enablePerformanceCheck',
      `
    <template>
      <div>{{ val }}</div>
    </template>
      `,
      'page',
      'page.ux',
      {
        compileOptionsObject: {
          // 删除 enablePerformanceCheck 相关，此处仅保留 compileOptionsObject 使用示例
          // enablePerformanceCheck: true
        }
      }
    ],
    [
      'without <template /> tag',
      `
      <div> without template tag </div>
      `,
      'page',
      'page.ux'
    ]
  ]
  beforeEach(() => {
    Object.assign(mockLoader, {
      emitError: jest.fn(),
      emitWarning: jest.fn()
    })
  })

  // exports 扩展的参数
  it.each(samples)('%s', async (_, source, type, fileName, extopts = {}) => {
    let resolver
    const promise = new Promise((resolve, reject) => {
      resolver = resolve
    })

    // if (extopts.compileOptionsObject && extopts.compileOptionsObject.enablePerformanceCheck) {
    //   compileOptionsObject.enablePerformanceCheck =
    //     extopts.compileOptionsObject.enablePerformanceCheck
    // }
    const loader = Object.assign({}, mockLoader, {
      resourcePath: path.resolve(__dirname, fileName),
      resourceQuery: `?uxType=${type}`,
      callback(err, result) {
        resolver([err, result])
      },
      async: function () {
        return this.callback
      }
    })

    uxLoader.call(loader, source)
    const [error, result] = await promise
    // compileOptionsObject.enablePerformanceCheck = false

    expect(error).toMatchSnapshot('error')
    expect(rowify([source, result])).toMatchSnapshot('source-vs-result')
    expect(loader.emitError.mock.calls).toMatchSnapshot('emitted errors')
    expect(loader.emitWarning.mock.calls).toMatchSnapshot('emitted warnings')
  })
})
