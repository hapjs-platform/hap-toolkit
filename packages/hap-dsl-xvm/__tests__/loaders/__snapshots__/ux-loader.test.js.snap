// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ux-loader normal page with style: emitted errors 1`] = `Array []`;

exports[`ux-loader normal page with style: emitted warnings 1`] = `Array []`;

exports[`ux-loader normal page with style: error 1`] = `null`;

exports[`ux-loader normal page with style: source-vs-result 1`] = `
"
      <template>
        <div>{{ val }}</div>
      </template>
      <style>
      div {
        color: blue;
      }
      </style>
      
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

var $app_script$ = null
$app_define$('@app-component/page', [], function($app_require$, $app_exports$, $app_module$) {
    $app_module$.exports.template = require(\\"!../../lib/loaders/template-loader.js!../../lib/loaders/fragment-loader.js?index=0&type=template!./page.ux?uxType=page&\\")
    $app_module$.exports.style = require(\\"!../../lib/loaders/style-loader.js?index=0&type=style!../../lib/loaders/fragment-loader.js?index=0&type=style!./page.ux?uxType=page\\")
});
$app_bootstrap$('@app-component/page',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION });"
`;

exports[`ux-loader normal page, enablePerformanceCheck: emitted errors 1`] = `Array []`;

exports[`ux-loader normal page, enablePerformanceCheck: emitted warnings 1`] = `Array []`;

exports[`ux-loader normal page, enablePerformanceCheck: error 1`] = `null`;

exports[`ux-loader normal page, enablePerformanceCheck: source-vs-result 1`] = `
"
    <template>
      <div>{{ val }}</div>
    </template>
      
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

var $app_script$ = null
$app_define$('@app-component/page', [], function($app_require$, $app_exports$, $app_module$) {
    $app_module$.exports.template = require(\\"!../../lib/loaders/template-loader.js!../../lib/loaders/fragment-loader.js?index=0&type=template!./page.ux?uxType=page&\\")
});
$app_bootstrap$('@app-component/page',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION });"
`;

exports[`ux-loader normal page: emitted errors 1`] = `Array []`;

exports[`ux-loader normal page: emitted warnings 1`] = `Array []`;

exports[`ux-loader normal page: error 1`] = `null`;

exports[`ux-loader normal page: source-vs-result 1`] = `
"
    <template>
      <div>{{ val }}</div>
    </template>
      
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

var $app_script$ = null
$app_define$('@app-component/page', [], function($app_require$, $app_exports$, $app_module$) {
    $app_module$.exports.template = require(\\"!../../lib/loaders/template-loader.js!../../lib/loaders/fragment-loader.js?index=0&type=template!./page.ux?uxType=page&\\")
});
$app_bootstrap$('@app-component/page',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION });"
`;

exports[`ux-loader without <template /> tag: emitted errors 1`] = `
Array [
  Array [
    [Error: 需要模板 <template> 片段],
  ],
]
`;

exports[`ux-loader without <template /> tag: emitted warnings 1`] = `Array []`;

exports[`ux-loader without <template /> tag: error 1`] = `null`;

exports[`ux-loader without <template /> tag: source-vs-result 1`] = `
"
      <div> without template tag </div>
      
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

var $app_script$ = null
$app_define$('@app-component/page', [], function($app_require$, $app_exports$, $app_module$) {
    $app_module$.exports.template = {}
});
$app_bootstrap$('@app-component/page',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION });"
`;
